src/digraph.rs

```diff
@@ -2265,6 +2278,141 @@ impl PyDiGraph {
+    /// :param weight_combo_fn: An optional python callable that, when
+    ///     specified, is used to merge parallel edges introduced by the
+    ///     contraction, which will occur when multiple nodes in
+    ///     ``nodes`` have an incoming edge
+    ///     from the same source node or when multiple nodes in
+    ///     ``nodes`` have an outgoing edge to the same target node.
+    ///     If this instance of :class:`~retworkx.PyDiGraph` is a multigraph,
+    ///     leave this unspecified to preserve parallel edges. If unspecified
+    ///     when not a multigraph, parallel edges and their weights will be
+    ///     combined by choosing one of the edge's weights arbitrarily based
+    ///     on an internal iteration order, subject to change.
+    /// :returns: The index of the newly created node.
+    /// :raises DAGWouldCycle: The cycle check is enabled and the
+    ///     contraction would introduce cycle(s).
+    #[pyo3(
+        text_signature = "(self, nodes, obj, /, check_cycle=None, weight_combo_fn=None)"
+    )]
+    pub fn contract_nodes(
+        &mut self,
+        py: Python,
+        nodes: Vec<usize>,
+        obj: PyObject,
+        check_cycle: Option<bool>,
+        weight_combo_fn: Option<PyObject>,
+    ) -> PyResult<usize> {
+        let can_contract = |nodes: &IndexSet<NodeIndex, ahash::RandomState>| {
+            // Start with successors of `nodes` that aren't in `nodes` itself.
+            let visit_next: Vec<NodeIndex> = nodes
+                .iter()
+                .flat_map(|n| self.graph.edges(*n))
+                .filter_map(|edge| {
+                    let target_node = edge.target();
+                    if !nodes.contains(&target_node) {
+                        Some(target_node)
+                    } else {
+                        None
+                    }
+                })
+                .collect();
+
+            // Now, if we can reach any of `nodes`, there exists a path from `nodes`
+            // back to `nodes` of length > 1, meaning contraction is disallowed.
+            let mut dfs = Dfs::from_parts(visit_next, self.graph.visit_map());
+            while let Some(node) = dfs.next(&self.graph) {
+                if nodes.contains(&node) {
+                    // we found a path back to `nodes`
+                    return false;
+                }
+            }
+            true
+        };
+
+        let mut indices_to_remove: IndexSet<NodeIndex, ahash::RandomState> =
+            nodes.into_iter().map(NodeIndex::new).collect();
+
+        if check_cycle.unwrap_or(self.check_cycle)
+            && !can_contract(&indices_to_remove)
+        {
+            return Err(DAGWouldCycle::new_err(
+                "Contraction would create cycle(s)",
+            ));
+        }
+
+        // Create new node.
+        let node_index = self.graph.add_node(obj);
+
+        // Sanitize new node index from user input.
+        indices_to_remove.remove(&node_index);
```

Comment:
Is this coming up somewhere? At this point I would never expect this to actually do anything since `node_index` was just created it would be kind of weird to have user input that includes it.

Reply:
Looking through the tests I understand now, you're guarding against someone passing in an invalid node that happens to end up being the new node. I don't think this will come up very often, but it should be low overhead as it's just a lookup so it's probably fine.

Reply:
Cool. Yeah that's right. We say that this method behaves like `add_node` when the user gives a list of nodes that is either completely empty or doesn't contain any nodes we know about. Adding this defensive removal ensures that is correct even for the case where the `nodes` parameter has an unknown node index that happens to become the index of the created node.